File: LeetCode-Python/BestTimetoBuyandSellStockIII.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        n = len(prices)
        left = [0] * n  # 0..i arası en iyi işlem
        right = [0] * (n + 1)  # i..n-1 arası en iyi işlem

        # Soldan sağa: ilk işlem
        min_price = prices[0]
        for i in range(1, n):
            min_price = min(min_price, prices[i])
            left[i] = max(left[i - 1], prices[i] - min_price)

        # Sağdan sola: ikinci işlem
        max_price = prices[-1]
        for i in range(n - 1, -1, -1):
            max_price = max(max_price, prices[i])
            right[i] = max(right[i + 1], max_price - prices[i])

        # İki işlemi birleştir
        max_profit = 0
        for i in range(n):
            max_profit = max(max_profit, left[i] + right[i + 1])

        return max_profit

File: LeetCode-Cpp/BestTimetoBuyandSellStockIII.cpp
if (prices.empty()) return 0;

        int n = prices.size();
        vector<int> left(n, 0);  // 0..i arası en iyi işlem
        vector<int> right(n + 1, 0); // i..n-1 arası en iyi işlem

        // Soldan sağa: ilk işlem
        int minPrice = prices[0];
        for (int i = 1; i < n; ++i) {
            minPrice = min(minPrice, prices[i]);
            left[i] = max(left[i - 1], prices[i] - minPrice);
        }

        // Sağdan sola: ikinci işlem
        int maxPrice = prices[n - 1];
        for (int i = n - 1; i >= 0; --i) {
            maxPrice = max(maxPrice, prices[i]);
            right[i] = max(right[i + 1], maxPrice - prices[i]);
        }

        // İki işlemi birleştir
        int maxProfit = 0;
        for (int i = 0; i < n; ++i) {
            maxProfit = max(maxProfit, left[i] + right[i + 1]);
        }

        return maxProfit;
